File: Data-Structures-and-Algorithms-using-Swift/Sorting/SelectionSort.swift
import Foundation

func greater(_ value1 : Int, _ value2 : Int) -> Bool {
    return value1 > value2
}

func selectionSort(_ arr : inout [Int]) {
    // sorted array created in reverse order.
    let size : Int = arr.count
    var mx : Int, j : Int, i : Int = 0
    while (i < size - 1) {
        mx = 0
        j = 1
        while (j < size - i) {
            if (arr[j] > arr[mx]) {
                mx = j
            }
            j += 1
        }
        arr.swapAt(size-1-i, mx)
        i += 1
    }
}

func selectionSort2(_ arr : inout [Int]) {
    // sorted array created in forward direction
    let size : Int = arr.count
    var mn : Int, j : Int, i : Int = 0
    while (i < size - 1) {
        mn = i
        j = i + 1
        while (j < size) {
            if (arr[j] < arr[mn]) {
                mn = j
            }
            j += 1
        }
        arr.swapAt(i, mn)
        i += 1
    }
}

var array : [Int] = [9, 1, 8, 2, 7, 3, 6, 4, 5]
selectionSort(&array)
print(array)
var array2 : [Int] = [9, 1, 8, 2, 7, 3, 6, 4, 5]
selectionSort2(&array2)
print(array2)

/*
[1, 2, 3, 4, 5, 6, 7, 8, 9]
[1, 2, 3, 4, 5, 6, 7, 8, 9]
*/

File: Data-Structures-Algorithms-In-TS/Sorting/SelectionSort.ts
function more(value1: number, value2: number): boolean {
    return value1 > value2;
}

function SelectionSort(arr: Array<number>) {
    let size: number = arr.length;
    let i: number;
    let j: number;
    let max: number;
    let temp: number;
    for (i = 0; i < size - 1; i++) {
        max = 0;
        for (j = 1; j < size - i; j++) {
            if (arr[j] > arr[max]) {
                max = j;
            }
        }
        temp = arr[size - 1 - i];
        arr[size - 1 - i] = arr[max];
        arr[max] = temp;
    }
}

function SelectionSort2(arr: Array<number>) {
    let size: number = arr.length;
    let i: number;
    let j: number;
    let min: number;
    let temp: number;
    for (i = 0; i < size - 1; i++) {
        min = i;
        for (j = i + 1; j < size; j++) {
            if (arr[j] < arr[min]) {
                min = j;
            }
        }
        temp = arr[i];
        arr[i] = arr[min];
        arr[min] = temp;
    }
}

// Testing code.
let array: Array<number> = [9, 1, 8, 2, 7, 3, 6, 4, 5];
SelectionSort(array);
console.log(array);

console.log();
let array2: Array<number> = [9, 1, 8, 2, 7, 3, 6, 4, 5];
SelectionSort2(array2);
console.log(array2);

